### ISSUE TITLE
Recomplete courses

### ISSUE THREAD

**Issue by tbeachy** [iomad/iomad#1561]
Trying to get a sub-set of our courses set up for recompletion using the [Course Recompletion plugin](https://moodle.org/plugins/local_recompletion) , and I'm a little confused about how this will work with Iomad 3.9.
First of all, it doesn't seem that there is any recomplete functionality built into Iomad, other than setting an expiration date on a course.  If that's incorrect, please let me know. 
My main question, however, is whether local_iomad_track will update with a new completion record, each time a student re-completes a course?  In other words, will new completions on a previously completed course be flagged and write a new record to local_iomad_track each time?  Due to the way to the recomplete plugin works, the previous completion gets written to a separate table, and the course_completions record gets updated.  So the only way to detect this in Iomad would be if it synced up every cron run and wrote new records when updated data exists.
It doesn't appear to in my testing, but perhaps I'm missing something.  
Thanks for any information.

**Comment by turf212** [comment]
Iomad will handle re-certification of courses through it's own processes without the need for additional plugins.  The workflow is as follows:

User is enrolled onto a course / provided a license to a course
- Entry is created into the local_iomad_track table for this
User undergoes activities in the couse
User completes the course (using standard Moodle course completion rules)
- Entry is updated in the local_iomad_track table with final grade and completion dates
- If there are Iomad certificates in the course then these are stored
- If the course has a value in the Iomad Course settings for training expires after a timestamp is written to the local_iomad_track table for when the user should be automatically cleared out of the course
Completion record is available in the Iomad reports and either:
1 - admin/manager manually clears the user from the current enrolment, or
2 - time is reached for user to be expired from the course 
which causes the user to be unenrolled, license marked as completed with, all standard Moodle activity interactions and grades are wiped.
User is re-enrolled in the course, creating a new entry in local_iomad_track and the cycle repeats.

We use the enrolment timestamp and course and user for uniqueness so, if a user is currently enrolled on a course and an admin clears and changes the completion criteria, then if the user already has a completion timestamp for their enrolment timestamp, nothing else is done.

In the latest 3.6 - 3.9 branches there has been functionality to upload previous course record information.  There is also the option to turn on editing when viewing a user's completion report where you can change the values stored, remove the entry completely or manually add in a single new entry.

FYI we use events to keep the records up to date rather than anything through cron specifically (e.g. user enrolled into  a course, user allocated a license, user used a license, user graded, user completed course and so on).

TLDR: the course expires after is used to allow you to perform automated clear-down and re-certification of users within courses.

**Comment by tbeachy** [comment]
That is excellent news, thank you for the detailed explanation Derick!
I will set it up and test it out.
Couple follow-up questions:

1.  The act of un-enrol/removing enrolment from a course is what causes activity completion records to be cleared out?
2.  I have not seen the "upload previous course record information" area yet.  Where would I access that in 3.9?

Thanks again!

**Comment by turf212** [comment]
1 - The user details are only cleared down from the course when the "Clear" button is used against that record in the User's course completion report page *or* when the user is cleared out automatically when they hit their time expires.
2 - It's probably the last commit to the code base.  If you go to site administration -> Iomad -> Import completion records, there are two options - from the current Moodle completion tables (so for a migration) and from a CSV upload.  I still need to add an example file but the minimum columns are:

username,coursename,timecompleted
user52712,course11,"2021/03/02"

If the course is a licensed course then you need to add a licensename header.  The complete list of headers available are: username', 'userid', 'courseid', 'coursename', 'timeenrolled', 'timestarted', 'timecompleted',
        'finalscore', 'licensename', 'licenseallocated', 'licenseid', 'companyid', 'company', 'departmentid', 'department'

It works for current courses only as there is sanity checking.  So if you have old courses which are not on this system then, for the moment, you need to create them.  I can't think of a better way for that to work with the current report requirements.

**Comment by tbeachy** [comment]
Ok, that's helpful.  Thanks again!

**Comment by tbeachy** [comment]
I've set up a test of this, and it appears that the automatic clear-out, based on expiration date, is not working.
I have a test user that completed a course on 3/11 and iomad certificate was issued.  Expiration date is 3/12.  It's now 3/15 and all activity completion etc. is still there.  I've gone through your instructions multiple times and can't see where I've gone wrong.
What am I missing?
![image](https://user-images.githubusercontent.com/5694016/111186892-75e2b800-8581-11eb-8b90-8c63a7b90827.png)

**Comment by tbeachy** [comment]
Looks like this may be the same issue: https://github.com/iomad/iomad/issues/1558

I previously turned off all iomad emails for the company in question, but have since turned some of the relevant ones for course expiration back on to see if that makes a difference.  So far it hasn't, and I've checked that appropriate cron jobs are running.

**Comment by tbeachy** [comment]
After more investigation, it seems that entering a value in the "Enrolment expires after (days)" field is what triggers the course to clear out activity completion and enrollment... which makes sense.  It's confusing because the help field for "Training expires after" says "The user's enrolment will be removed at the time of expiry".

**Comment by tbeachy** [comment]
Closing.

**Comment by karthikashekar7** [comment]
@turf212 @tbeachy Thank you ,  This helped me alot .I have a question, do i need to set "Enrolment expires after (days)" and "Training expires after (days)" to clear out the course completion data? I have this question because if i do not set Training expires after (days)" then in mdl_local_iomad_track timeexpires is null. I would also like to know which scheduled task will trigger this course completion to clear as i tried running "local_email_reports" after changing timeexpires as same as timecompeleted but it did not clear the course completion record ?

**Comment by tbeachy** [comment]
As far as I know, you only have to set "Enrolment expires after (days)" to clear out completion data.  The training expiration date is what shows up in the Dashboard, but I don't know if it triggers any functionality or not.
Try running the local_iomad\task\cron_task  to clear enrolments...

**Comment by turf212** [comment]
The process is enabled by setting a value for "Training expires after".  The user is cleared down from the course at the earliest of:

- The value after completion of training expires after.
- The value after completion of training expires after minus days to send email before.
- the value of "enrolment expires after" after the course is completed by the user.

The task which deals with this is, confusingly, the local/email_reports task which expires users once a day.  If you are testing with a value of 1 then, depending on when the user completed the course, it could be 47 hours before they are actually removed.

**Comment by karthikashekar7** [comment]
@turf212 @tbeachy  Thank you, it works  after setting a value for  "Training expires after (days)"

**Comment by abdallah84** [comment]
Hi @turf212 ,
my version 3.8
I have one quiz in course with unlimited attempts , how can let moodle/iomad automatically take the highest attempts after the course is completed and update mdl_local_iomad_track ?
